Tests/OpenCombineTests/PublisherTests/FlatMapTests.swift

```diff
@@ -0,0 +1,539 @@
+
+        let downstreamSubscriber = TrackingSubscriber(receiveSubscription: {
+            $0.request(.unlimited)
+        })
+
+        flatMap.subscribe(downstreamSubscriber)
+
+        upstreamPublisher.send(childPublisher1)
+        upstreamPublisher.send(childPublisher2)
+
+        childPublisher1.send(666)
+        childPublisher2.send(777)
+        XCTAssertEqual(downstreamSubscriber.history, [.subscription("FlatMap"),
+                                                      .value(666),
+                                                      .value(777)])
+    }
+
+    // This test ensures that the code can properly re-enter when synchronously receiving
+    // a value during subscription (which Publishers.Just does).
+    //  1. FlatMap.Inner.receive(_ input:)
+    //    2. Publisher.subscribe
+    //      ...
+    //      3. FlatMap.Inner.ChildSubscriber.recive(subscription:)
+    //        4. subscription.request()
+    //          5. Just.Inner.request()
+    //            6. FlatMap.Inner.child(_:receivedValue)
+    //              7. lock
+    //
+    // At one point, I had a bug where the lock was taken by #1 before calling #2
+    // This broke the rules of calling out with a lock held, and lead to a deadlock
+    // at #7.
+    //
+    // Also, in my opinion, working around the issue with recursive locks is a smell.
+    func testChildSubscribeDeadlock() {
+        let upstreamSubscription = CustomSubscription()
+        let upstreamPublisher = CustomPublisherBase<Int, Never>(
+            subscription: upstreamSubscription)
+
+        let flatMap = upstreamPublisher.flatMap(maxPublishers: .max(2)) { Just($0) }
+
+        let downstreamSubscriber = TrackingSubscriberBase<Int, Never>(
+            receiveSubscription: { $0.request(.unlimited) })
+
+        flatMap.subscribe(downstreamSubscriber)
+        XCTAssertEqual(upstreamPublisher.send(666), .none)
+
+        // Simply making it here shows that there's no dealock
+    }
+
+    func testCancelCancels() {
+        let upstreamSubscription = CustomSubscription()
+        let upstreamPublisher = CustomPublisherBase<Int, Never>(
+            subscription: upstreamSubscription)
+
+        let childSubscription = CustomSubscription()
+        let childPublisher = CustomPublisherBase<Int, Never>(
+            subscription: childSubscription)
+
+        let flatMap = upstreamPublisher.flatMap { _ in childPublisher }
+
+        var downstreamSubscription: Subscription?
+        let downstreamSubscriber = TrackingSubscriberBase<Int, Never>(receiveSubscription:
+        {
+            downstreamSubscription = $0
+            $0.request(.unlimited)
+        })
+
+        flatMap.subscribe(downstreamSubscriber)
+
+        XCTAssertEqual(upstreamPublisher.send(1), .none)
+
+        downstreamSubscription?.cancel()
+
+        XCTAssertEqual(upstreamSubscription.history.last!, .cancelled)
+        XCTAssertEqual(childSubscription.history.last!, .cancelled)
```

Comment:
`Optional` conditionally conforms to `Equatable`, so there's no need to force-unwrap here. We don't want to crash in tests, we want to fail gracefully.